src/asm_files.cpp

```diff
@@ -250,8 +248,27 @@ static void append_subprograms(raw_program& prog, const vector<raw_program>& pro
                                      std::to_string(reloc.source_offset));
             }
             ELFIO::section& subprogram_section = *reader.sections[section_index];
-            auto subprogram = read_subprogram(subprogram_section, symbols, symbol_name);
-            prog.prog.insert(prog.prog.end(), subprogram.begin(), subprogram.end());
+
+            auto subprogram = find_subprogram(programs, subprogram_section, symbol_name);
+            if (!subprogram) {
+                // The program will be invalid, but continue rather than throwing an exception
+                // since we might be verifying a different program in the file.
+                return std::string("Subprogram '" + symbol_name + "' not found in section '" +
+                                   subprogram_section.get_name() + "'");
+            }
+            auto& subprog = subprogram->get();
+
+            // Make sure subprogram has already had any subprograms of its own appended.
+            std::string error = append_subprograms(subprog, programs, function_relocations, reader, symbols);
+            if (!error.empty()) {
+                return error;
+            }
+
+            // Append subprogram to program.
+            prog.prog.insert(prog.prog.end(), subprog.prog.begin(), subprog.prog.end());
+            for (int i = 0; i < subprog.info.line_info.size(); i++) {
+                prog.info.line_info[prog.info.line_info.size()] = subprog.info.line_info[i];
+            }
```

Comment:
The suggestion looks like incorrect logic to me, since prog.size() and line_info.size() should always end up the same value in a correct implementation.  The existing code is correct in that sense and is easy to see that it will be.